app/src/main/java/com/ecchilon/sadpanda/api/DataLoader.java

```diff
@@ -51,252 +52,288 @@
 	private static final String PHOTO_URL_EX = "http://exhentai.org/s/%s/%d-%d";
 	private static final String GALLERY_PATTERN = "http://(g\\.e-|ex)hentai\\.org/g/(\\d+)/(\\w+)/";
 
-	private static final Pattern pPhotoUrl = Pattern.compile("width:(\\d+)px; height:(\\d+)px; background:transparent url\\((.+?)\\) -(\\d+)px 0 no-repeat\"><a href=\"http://exhentai\\.org/s/(\\w+?)/\\d+-(\\d+)");
+	private static final MediaType JSON = MediaType.parse("application/json; charset=utf-8");
+
+	private static final Pattern pPhotoUrl = Pattern.compile(
+			"width:(\\d+)px; height:(\\d+)px; background:transparent url\\((.+?)\\) -(\\d+)px 0 no-repeat\"><a "
+					+ "href=\"http://exhentai\\.org/s/(\\w+?)/\\d+-(\\d+)");
 	private static final Pattern pShowkey = Pattern.compile("var showkey.*=.*\"([\\w-]+?)\";");
 	private static final Pattern pImageSrc = Pattern.compile("<img id=\"img\" src=\"(.+)/(.+?)\"");
 	private static final Pattern pGalleryHref = Pattern.compile("<a href=\"" + GALLERY_PATTERN + "\" onmouseover");
 	private static final Pattern pGalleryUrl = Pattern.compile(GALLERY_PATTERN);
 
-	private final HttpClient mHttpClient;
-	private final HttpContext mHttpContext;
+	private final OkHttpClient client;
+	private final ExhentaiAuth auth;
 
 	@Inject
-	private DataLoader(HttpClient httpClient, HttpContext context) {
-		mHttpClient = httpClient;
-		mHttpContext = context;
-	}
-
-	private HttpResponse getHttpResponse(HttpRequestBase httpRequest) throws IOException {
-		return mHttpClient.execute(httpRequest, mHttpContext);
+	DataLoader(OkHttpClient client, ExhentaiAuth auth) {
+		this.client = client;
+		this.auth = auth;
 	}
 
-	private JSONObject callApi(JSONObject json) throws ApiCallException {
-		assertNotMainThread();
-		String responseStr = "";
-
-		try {
-			HttpPost httpPost = new HttpPost(API_URL_EX);
-
-			httpPost.setHeader("Accept", "application/json");
-			httpPost.setHeader("Content-Type", "application/json");
-			httpPost.setEntity(new StringEntity(json.toString()));
+	private Observable<JSONObject> callApi(JSONObject json) {
+		return Observable.<JSONObject>create(subscriber -> {
+			Request request = new Request.Builder()
+					.addHeader("Accept", "application/json")
+					.addHeader("Cookie", auth.getSessionCookie())
+					.url(API_URL_EX)
+					.post(RequestBody.create(JSON, json.toString()))
+					.build();
+
+			Response response;
+			String responseStr;
+			try {
+				response = client.newCall(request).execute();
+				responseStr = response.body().string();
+			}
+			catch (IOException e) {
+				if (!subscriber.isUnsubscribed()) {
+					subscriber.onError(OnErrorThrowable.from(new ApiCallException(ApiErrorCode.IO_ERROR, e)));
```

Comment:
Consider dropping the `ApiCallException`. It's rather redundant with the way observables handle errors automatically.